ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/LfValueTranslation.scala

```diff
@@ -0,0 +1,220 @@
+    LfEngineToApi.assertOrRuntimeEx(
+      failureContext = s"attempting to deserialize persisted $attribute to value",
+      LfEngineToApi
+        .lfVersionedValueToApiValue(
+          verbose = verbose,
+          value = value,
+        ),
+    )
+
+  private def toApiRecord(
+      value: LfValue,
+      verbose: Boolean,
+      attribute: => String,
+  ): ApiRecord =
+    LfEngineToApi.assertOrRuntimeEx(
+      failureContext = s"attempting to deserialize persisted $attribute to record",
+      LfEngineToApi
+        .lfVersionedValueToApiRecord(
+          verbose = verbose,
+          recordValue = value,
+        ),
+    )
+
+  def deserialize[E](raw: Raw.Created[E], verbose: Boolean): CreatedEvent = {
+    val key = LfValueTranslation.Cache.Key(raw.partial.eventId)
+    val create =
+      cache
+        .getIfPresent(key)
+        .getOrElse(
+          LfValueTranslation.Cache.Value.Create(
+            argument = ValueSerializer.deserializeValue(raw.createArgument),
+            key = raw.createKeyValue.map(ValueSerializer.deserializeValue)
+          )
+        )
+        .assertCreate()
+    raw.partial.copy(
+      createArguments = Some(
+        toApiRecord(
+          value = create.argument,
+          verbose = verbose,
+          attribute = "create argument",
+        )
+      ),
+      contractKey = create.key.map(
+        key =>
+          toApiValue(
+            value = key,
+            verbose = verbose,
+            attribute = "create key",
+        )
+      ),
+    )
+  }
+
+  def deserialize(raw: Raw.TreeEvent.Exercised, verbose: Boolean): ExercisedEvent = {
+    val key = LfValueTranslation.Cache.Key(raw.partial.eventId)
+    val exercise =
+      cache
+        .getIfPresent(key)
+        .getOrElse(
+          LfValueTranslation.Cache.Value.Exercise(
+            argument = ValueSerializer.deserializeValue(raw.exerciseArgument),
+            result = raw.exerciseResult.map(ValueSerializer.deserializeValue)
+          )
+        )
+        .assertExercise()
+    raw.partial.copy(
+      choiceArgument = Some(
+        toApiValue(
+          value = exercise.argument,
+          verbose = verbose,
+          attribute = "exercise argument",
+        )
+      ),
+      exerciseResult = exercise.result.map(
+        result =>
+          toApiValue(
+            value = result,
+            verbose = verbose,
+            attribute = "exercise result",
+        )
+      ),
+    )
+  }
+
+}
+
+object LfValueTranslation {
+
+  type Cache = caching.Cache[Cache.Key, Cache.Value]
+
+  object Cache {
+
+    private implicit object `Key Weight` extends caching.Weight[Key] {
+      override def weigh(value: Key): caching.Cache.Size =
+        0 // make sure that only the value is counted
+    }
+
+    private implicit object `Value Weight` extends caching.Weight[Value] {
+      override def weigh(value: Value): caching.Cache.Size =
+        1 // TODO replace this with something to avoid weights entirely
```

Comment:
Just construct a count-based caffeine cache, i.e., use `.maximumSize(...)` instead of `.maximumWeight(...)` and a weighter when creating the cache:
https://github.com/ben-manes/caffeine/blob/4451d4443f1db71d3dc11925a1bca9a5206dd539/caffeine/src/main/java/com/github/benmanes/caffeine/cache/Caffeine.java#L368

Reply:
I tried this already, unfortunately this starts to get out of scope with several changes needed to make sure we have an alternative that doesn't take an implicit `Weight` and that everything typechecks. I agree it's a meaningful improvement but I'd prefer not making everything in one PR which is already reasonably large.